src/plugins_types.c

```diff
@@ -1073,9 +1091,38 @@ lyplg_type_resolve_leafref(const struct lysc_type_leafref *lref, const struct ly
     if (target) {
         *target = set.val.nodes[i].node;
     }
+    if (targets) {
+        for (i = 0; i < set.used; ++i) {
+            if (set.val.nodes[i].type != LYXP_NODE_ELEM) {
+                continue;
+            }
+            if (((struct lyd_node_term *)set.val.nodes[i].node)->value.realtype != value->realtype) {
+                continue;
+            }
+
+            if (!lref->plugin->compare(&((struct lyd_node_term *)set.val.nodes[i].node)->value, value)) {
+                LY_ARRAY_NEW_GOTO(LYD_CTX(node), *targets, item, rc, cleanup);
+                *item = set.val.nodes[i].node;
+            }
+        }
+    }
 
 cleanup:
     lyxp_expr_free(LYD_CTX(node), target_path);
     lyxp_set_free_content(&set);
     return rc;
 }
+
+LIBYANG_API_DEF LY_ERR
+lyplg_type_resolve_leafref(const struct lysc_type_leafref *lref, const struct lyd_node *node, struct lyd_value *value,
+        const struct lyd_node *tree, struct lyd_node **target, char **errmsg)
+{
+    return _lyplg_type_resolve_leafref(lref, node, value, tree, target, NULL, errmsg);
+}
+
+LIBYANG_API_DEF LY_ERR
+lyplg_type_resolve_leafref2(const struct lysc_type_leafref *lref, const struct lyd_node *node, struct lyd_value *value,
```

Comment:
I realized another thing, you are returning a LY_ARRAY from an API function, I do not think we ever do that and return `ly_set` instead, so please do that. I suppose using LY_ARRAY in the hash table records makes sense because it is a bit smaller (but maybe not worth it) so it is up to you to design it sensibly.

Reply:
Done